### PR TITLE
Move off of Respect\Validation

### PR DESCRIPTION
**Pull request by oschwald**
Also, update `with*` methods to support typed named arguments.

### CODE REVIEW

File: src/MinFraud.php [hunk 1]
```diff
@@ -113,82 +113,366 @@ public function __construct(
...
-        $new->content = $values;
+        $this->verifyEmpty($values);
 
         return $new;
     }
 
     /**
      * This returns a `MinFraud` object with the `device` array set to
-     * `$values`. Existing `device` data will be replaced.
+     * to the values provided. Existing `device` data will be replaced.
```
Comment (ugexe) [inline src/MinFraud.php:160]: `s/* to /* /`here any for many of the other function comments

File: src/MinFraud.php [hunk 2]
```diff
@@ -113,82 +113,366 @@ public function __construct(
...
+     * @param array       $values         An array of device data. The keys are the same as
+     *                                    the JSON keys. You may use either this or the named
+     *                                    arguments, but not both.
+     * @param string|null $ipAddress      The IP address associated with the device
+     *                                    used by the customer in the transaction.
+     *                                    The IP address must be in IPv4 or IPv6
+     *                                    presentation format, i.e., dotted-quad
+     *                                    notation or the IPv6 hexadecimal-colon
+     *                                    notation.
+     * @param string|null $userAgent      the HTTP `User-Agent` header of the browser
```
Comment (ugexe) [inline src/MinFraud.php:171]: `s/the HTTP/The HTTP/`. A couple other descriptions also start with a lower case letter

File: src/MinFraud.php [hunk 3]
```diff
@@ -198,59 +482,631 @@ public function withEmail(array $values): self
...
+        if ($phoneNumber !== null) {
+            $values['phone_number'] = $phoneNumber;
+        }
+
+        if ($postal !== null) {
+            $values['postal'] = $postal;
+        }
+
+        if ($region !== null) {
+            if (!preg_match('/^[0-9A-Z]{1,4}$/', $region)) {
```
Comment (ugexe) [inline src/MinFraud.php:599]: I see `{1,4}` matches the original validation rule (and what [the docs](https://dev.maxmind.com/minfraud/api-documentation/requests/#schema--request--billing__region) say is the max length), but shouldn't this be `{1,3}`? Similarly to the other validation on `$region` in `withShipping`. Although I also see (and the max length mentioned by [the docs

File: src/MinFraud.php [hunk 4]
```diff
@@ -198,59 +482,631 @@ public function withEmail(array $values): self
...
+                $this->maybeThrowInvalidInputException(
+                    "$region is not a valid ISO 3166-2 region code (without country prefix)",
```
Comment (ugexe) [inline src/MinFraud.php:601]: In the `withShipping` validation error message it doesn't mention the `(without country prefix)`. I suppose they should both be the same one way or the other.

**Review by ugexe — CHANGES_REQUESTED** [review]
The changes looks accurate to me, although I do have a question about how we've already been handling the region codes.

File: src/MinFraud.php (hunk 3 above)
Comment (oschwald) [reply src/MinFraud.php:599]: Yeah, I was wondering about the 4 in the existing validation too but didn't dig into it. I'll double check what the web service does.

Comment (oschwald) [reply src/MinFraud.php:599]: It appears the web service uses `\A[0-9A-Z]{1,4}\z`. That is probably something worth looking into, but I am inclined to leave it for now.

**Comment by oschwald** [comment]
Thanks! I made the changes and rebased to fix the `CHANGELOG.md` confict.

**Review by ugexe — APPROVED** [review]

**Pull request merged**